### PR TITLE
JSpecify: handle return types of method references in Java Generics

### PR DESCRIPTION
**Pull request by akulk022**
Adding new test case and the logic to handle the negative scenario of the test case where we were getting a false positive earlier.

```Java
class Test {
    interface A<T1 extends @Nullable Object> {
        T1 function(Object o);
    }
    static @Nullable String foo(Object o) {
        return o.toString();
    }
    static void testPositive() {
    // BUG: Diagnostic contains: referenced method returns @Nullable
	A<String> p = Test::foo;
    }
    static void testNegative() {
    // We were getting a false positive error on this line as the Nullability of the overriden method was not identified
        A<@Nullable String> p = Test::foo;
    }
}
```

All test cases in NullAwayJSpecifyGenericTests.java passed for these changes.

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java [hunk 1]
```diff
@@ -735,8 +735,13 @@ private static Type getTypeForSymbol(Symbol symbol, VisitorState state) {
...
   }
 
-  private static Nullness getGenericMethodReturnTypeNullness(
-      Symbol.MethodSymbol method, Type enclosingType, VisitorState state, Config config) {
+  static Nullness getGenericMethodReturnTypeNullness(
+      Symbol.MethodSymbol method, @Nullable Type enclosingType, VisitorState state, Config config) {
+    if (enclosingType == null) {
+      // we have no additional information from generics, so return NONNULL (presence of a @Nullable
+      // annotation should have been handled by the caller)
+      return Nullness.NONNULL;
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:743]: I think you'll fix the code coverage issue here if you remove the similar code at lines 702-706 (which is no longer needed)

Comment (akulk022) [reply nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java:743]: Makes a lot of sense! Thanks!

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 2]
```diff
@@ -1003,9 +1007,14 @@ private boolean overriddenMethodReturnsNonNull(
...
-              overriddenMethod, enclosingSymbol, state, config)
-          .equals(Nullness.NONNULL);
+
+      return (memberReferenceTree != null)
+          ? GenericsChecks.getGenericMethodReturnTypeNullness(
+                  overriddenMethod, ASTHelpers.getType(memberReferenceTree), state, config)
+              .equals(Nullness.NONNULL)
+          : GenericsChecks.getGenericMethodReturnTypeNullness(
+                  overriddenMethod, enclosingSymbol, state, config)
+              .equals(Nullness.NONNULL);
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/NullAway.java:1017]: Can we rewrite this as an if-then-else similar to what we did in #845, with a comment to document each case?

File: nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java [hunk 3]
```diff
@@ -439,6 +439,31 @@ public void testForMethodReferenceInAnAssignment() {
         .doTest();
   }
 
+  @Test
+  public void testForMethodReferenceReturnType() {
```
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java:443]: As noted in an offline discussion, this test looks good but let's add tests for more (pseudo-)assignments, like passing a method reference as a function parameter and returning a method reference

**Review by msridhar — COMMENTED** [review]
Great start!  Couple comments

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 2 above)
Comment (akulk022) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java:1017]: Yes, Will do!

**Review by msridhar — APPROVED** [review]
LGTM!
